### ISSUE TITLE
FileLock

### ISSUE THREAD

**Issue by vemv** [ruby-concurrency/concurrent-ruby#555]
Hi!

I think I have the need for a File-based lock, because multiple web processes (as typical in Puma/Unicorn) can attempt to perform an overlapping operation. It only overlaps within a given machine - not across the fleet (so a distributed lock isn't the answer in this case).

I see a couple gems providing file locks, one of them looking pretty well-maintained.

I'm still hesitant to use any of those though - File-based locking is apparently a difficult technique and I'd expect its implementation/testing to be of the highest quality - today and tomorrow.

concurrent-ruby is a popular/reputed project which probably can give that degree of confidence.

Would you be willing to offer a File-based locking mechanism? Is there any technical concern?
## API

Personally I'd expect the API to be as simple as possible, and as similar as possible to a Mutex. I shouldn't know anything about unixy things to use it, nor there should be low-level flags, etc.

Capablities would include:
- lock
- unlock
- try-lock

I don't know if something akin to a ReadWriteLock should be provided. Personally I don't have that use case.

What do you think?

Cheers - Victor

**Comment by jdantonio** [comment]
@vemv Thank you very much for this suggestion. I've been giving it some thought and I'm very intrigued, but it's not something we could work on until after we've released 2.0. That's our highest priority. We are always open to PRs with new features for Edge so we're happy to support anyone in the community who would like to take this on, but @pitr-ch and I need to focus on 2.0 right now.

I'm going to tag this `paused` for now and we'll revisit after we've released 2.0.
